### ISSUE TITLE
no-return-await must be aware of try/catch/finally

### ISSUE THREAD

**Issue by daltones** [eslint/eslint#7581]
`no-return-await` currently doesn't consider if the `return` statement is inside a `try` block. In this case, the presence of `await` does meter.

Consider this:
```js
function asyncLog() {
  return new Promise((resolve) => {
    setTimeout(() => {
      console.log('first log');
      resolve(true);
    }, 1000);
  });
}

async function foo() {
  try {
    return await asyncLog();
    // I really need this await, but no-return-await is complaining about it :(
  } finally {
    console.log('second log');
  }
}

foo();
// first log
// second log


async function bar() {
  try {
    return asyncLog();
  } finally {
    console.log('second log');
  }
}

bar();
// second log
// first log


async function baz() {
  try {
    return asyncLog();
  } catch (error) {
    // if asyncLog() rejects, error will be swallowed because I missed await :(
  }
}
```

**Comment by not-an-aardvark** [comment]
Thanks, I can reproduce this. So if I'm understanding correctly, would the correct behavior be to not report any `return await` statements in `try` blocks?

**Comment by daltones** [comment]
@not-an-aardvark exactly. Inside `try`, you have to chose whether use or not, no warns at all.

**Comment by platinumazure** [comment]
What happens if you split the await from the return?

```js
try {
    const result = await asyncLog();
    return result;
} finally {
    // Do the thing
}
```

**Comment by not-an-aardvark** [comment]
@platinumazure That has the same behavior as

```js
try {
  return await asyncLog();
} finally {
    // Do the thing
}
```

**Comment by platinumazure** [comment]
Then I think we have a solution- no need for special logic here. Or am I
missing something? Why does the presence of a try/catch/finally block mean
we need to handle return await statements differently?

**Comment by daltones** [comment]
The problem is that the warning message of this rule is “Redundant use of `await` on a return value”.

The term “Redundant” is in fact incorrect in these `try` cases.

**Comment by not-an-aardvark** [comment]
The intention of the `no-return-await` rule is to flag unnecessary `return await` statements where `await` could be removed without observably changing the behavior of the code*. When `return await` is used within a `try` statement, this is not the case.

---

*aside from the fact that the Promise will fulfill one tick sooner
